src/solver/algorithms/Fallin_solver.py

```diff
@@ -0,0 +1,193 @@
+      for j in range(9):
+        if(data[i][j] != 0):
+          self.cells[i][j] = { data[i][j]: 1 }
+          self.count[i][j] = 1
+          self.known = self.known + 1
+        else:
+          self.cells[i][j] = {1:1,2:1,3:1,4:1,5:1,6:1,7:1,8:1,9:1}
+          self.count[i][j] = 9
+
+  # dumps to external data (ie, a simple matrix/array-of-arrays)
+  def dump(self):
+    out = []
+    for i in range(9):
+      out += [ [] ]
+      for j in range(9):
+        if(self.count[i][j] == 1):
+          out[i] += [self.cells[i][j].keys()[0]]
+        else:
+          out[i] += [0]
+    return out
+
+  def load_string(self, str):
+    block_of_nine = str.split("\n");
+    lines = []
+    for line in block_of_nine:
+      line = line.strip()
+      if(line == ''):
+        continue
+      row = []
+      block_of_three = line.split()
+      for field in block_of_three:
+        field = field.strip()
+        if(field == ''):
+          continue
+        element = int(field)
+        row += [element]
+      lines += [row]
+
+    self.load(lines)
+
+  def dump_string(self):
+    out = self.dump()
+    out_str = ""
+    for i in range(9):
+      for j in range(9):
+        out_str += str(out[i][j]) + " "
+        if(j % 3 == 2): out_str += "  "
+      out_str += "\n"
+      if(i % 3 == 2): out_str += "\n"
+    return out_str
+
+  # finds possible elements for each position
+  def reduce(self):
+    self.known = 0
+    # for each cell:
+    for i in range(9):
+      for j in range(9):
+        # skip this cell if it's already known
+        if(self.count[i][j] == 1):
+          self.known = self.known + 1
+          continue
+        # start with all possible
+        possible = {1: 1, 2: 1, 3: 1, 4: 1, 5: 1, 6: 1, 7: 1, 8: 1, 9: 1}
+        # check row
+        for n in range(9):
+          if(self.count[i][n] == 1):
+            x = self.cells[i][n].keys()[0]
+            if(possible.has_key(x)):
+              del possible[x]
+        # check column
+        for n in range(9):
+          if(self.count[n][j] == 1):
+            x = self.cells[n][j].keys()[0]
+            if(possible.has_key(x)):
+              del possible[x]
+        # check inner box
+        box_i = i - (i % 3)
+        box_j = j - (j % 3)
+        for a in range(box_i,box_i+3):
+          for b in range(box_j,box_j+3):
+            if(self.count[a][b] == 1):
+              x = self.cells[a][b].keys()[0]
+              if(possible.has_key(x)):
+                del possible[x]
+
+        # now we deal based on 'possible'
+        self.count[i][j] = len(possible.keys())
+
+        # none possible: no solutions
+        if(self.count[i][j] == 0):
+          return 0
+
+        # one possible: increase number of knowns
+        if(self.count[i][j] == 1):
+          self.known = self.known + 1
+
+        # store in 'cells' matrix
+        self.cells[i][j] = possible
+
+    # done: successful (ie, no impossible entries)
+    return 1
+
+  def solve(self, start = 0):
+
```

Comment:
same here method is bigger than 20 lines of code and also there are more than 2 nested levels